addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,448 @@
+// determined in the Reconcile function
+func (r *PackageInstallStatusReconciler) reconcile(ctx context.Context, clusterClient client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole, log logr.Logger) error {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cluster), clusterBootstrap); err != nil {
+		return err
+	}
+	clusterBootstrap = clusterBootstrap.DeepCopy()
+
+	// this shouldn't include kapp ctrl package as it'll get processed separately
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
+
+	var cbStatusChanged bool
+
+	for _, pkg := range packages {
+		if pkg == nil {
+			continue
+		}
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(ctx, clusterClient, &clusterBootstrap.Status, cluster.Name, clusterRole, pkg.RefName, constants.TKGSystemNS, log)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just continue with collecting PackageInstallStatus for other packages
+			cbStatusChanged = cbStatusChanged || r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, pkg.RefName)
+		}
+	}
+
+	// Note: kapp ctrl pkgi exists only for the workload cluster
+	// it is installed under cluster.Namespace in the management cluster and should be handled separately
+	if clusterRole == clusterRoleWorkload && clusterBootstrap.Spec.Kapp != nil {
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(ctx, r.Client, &clusterBootstrap.Status, cluster.Name, clusterRole, clusterBootstrap.Spec.Kapp.RefName, cluster.Namespace, log)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just proceed with updating the ClusterBootstrapStatus for other packages in one update operation
+			cbStatusChanged = cbStatusChanged || r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, clusterBootstrap.Spec.Kapp.RefName)
+		}
+	}
+
+	if !cbStatusChanged {
+		log.Info("no change in the reconciliation status of any of the core/additional packages in the cluster")
+		return nil
+	}
+
+	if err := r.Client.Status().Update(ctx, clusterBootstrap); err != nil {
```

Comment:
Nor sure if the patch helper should be used instead. 

This should make it unnecessary to track if the status changed.

Reply:
+1 to use patch, but given the object is not using CAPI conditions it is not possible to use the CAPI patch helper

Reply:
Interesting I thought I saw our patch helper used in other places in this repo.

Reply:
changed to use CAPI ```patchHelper.Patch()```

Reply:
We are doing that today in CAPI, but maybe that's not applicable here. I'll take a closer look

Reply:
@sbueringer I tried using cluster api util functions, but the ```Set``` function compares the  conditions by matching the types: https://github.com/kubernetes-sigs/cluster-api/blob/main/util/conditions/setter.go#L52

But as Carvel conditions are one of the followings in a mutual exclusive way ```{v1alpha1.Reconciling, v1alpha1.ReconcileFailed, v1alpha1.ReconcileSucceeded, v1alpha1.Deleting, v1alpha1.DeleteFailed}```, and I build the condition for a package by prefixing the package name with what I observe in Carvel, I need the prefix be compared not the full condition type name. That's why using CAPI's condition util, I get multiple different conditions for the same package which is unwanted as follows:(  As such, seems like I need to have my own comparison logic and can't use something like ```conditions.Set(clusterBootstrap, &condition)```

```
status:
  conditions:
  - lastTransitionTime: "2022-03-23T18:31:44Z"
    status: "True"
    type: Antrea-ReconcileSucceeded
  - lastTransitionTime: "2022-03-24T23:55:51Z"
    status: "True"
    type: Antrea-Reconciling
  - lastTransitionTime: "2022-03-24T02:10:20Z"
    status: "True"
    type: Kapp-Controller-ReconcileSucceeded
  - lastTransitionTime: "2022-03-23T18:31:47Z"
    status: "True"
    type: Metrics-Server-ReconcileSucceeded
  - lastTransitionTime: "2022-03-24T23:53:42Z"
    status: "True"
    type: Secretgen-Controller-ReconcileFailed
  - lastTransitionTime: "2022-03-24T02:10:17Z"
    status: "True"
    type: Secretgen-Controller-ReconcileSucceeded
  - lastTransitionTime: "2022-03-24T23:55:12Z"
    status: "True"
    type: Secretgen-Controller-Reconciling
  resolvedTKR: v1.22.5---vmware.1-tkg.3
```

Reply:
Makes sense!